cranelift/codegen/src/egraph.rs

```diff
@@ -1,342 +1,437 @@
-                    }
-                } else if has_side_effect(func, inst) || opcode.can_load() {
-                    self.stats.node_created += 1;
-                    self.stats.node_inst += 1;
-                    Node::Inst {
-                        op,
-                        args,
-                        ty,
-                        arity,
-                        srcloc,
-                        loop_level,
-                    }
-                } else {
-                    self.stats.node_created += 1;
-                    self.stats.node_pure += 1;
-                    Node::Pure {
-                        op,
-                        args,
-                        ty,
-                        arity,
-                    }
-                };
-                let dedup_needed = self.node_ctx.needs_dedup(&node);
-                let is_pure = matches!(node, Node::Pure { .. });
+    /// Remove pure nodes from the `Layout` of the function, ensuring
+    /// that only the "side-effect skeleton" remains, and also
+    /// optimize the pure nodes. This is the first step of
+    /// egraph-based processing and turns the pure CFG-based CLIF into
+    /// a CFG skeleton with a sea of (optimized) nodes tying it
+    /// together.
+    ///
+    /// As we walk through the code, we eagerly apply optimization
+    /// rules; at any given point we have a "latest version" of an
+    /// eclass of possible representations for a `Value` in the
+    /// original program, which is itself a `Value` at the root of a
+    /// union-tree. We keep a map from the original values to these
+    /// optimized values. When we encounter any instruction (pure or
+    /// side-effecting skeleton) we rewrite its arguments to capture
+    /// the "latest" optimized forms of these values. (We need to do
+    /// this as part of this pass, and not later using a finished map,
+    /// because the eclass can continue to be updated and we need to
+    /// only refer to its subset that exists at this stage, to
+    /// maintain acyclicity.)
+    fn remove_pure_and_optimize(&mut self) {
+        let mut cursor = FuncCursor::new(self.func);
+        let mut value_to_opt_value: SecondaryMap<Value, Value> =
+            SecondaryMap::with_default(Value::reserved_value());
+        let mut gvn_map: CtxHashMap<(Type, InstructionData), Value> =
+            CtxHashMap::with_capacity(cursor.func.dfg.num_values());
 
-                let mut id = self.egraph.add(node, &mut self.node_ctx);
+        // In domtree preorder, visit blocks. (TODO: factor out an
+        // iterator from this and elaborator.)
+        let root = self.domtree_children.root();
+        let mut block_stack = vec![root];
+        while let Some(block) = block_stack.pop() {
+            // We popped this block; push children
+            // immediately, then process this block.
+            for child in self.domtree_children.children(block) {
+                block_stack.push(child);
+            }
```

Comment:
This is shorter and allows for reserving capacity for all children at once:

```suggestion
            block_stack.extend(self.domtree_children.children(block));
```